examples/java/src/com/mlang/MValue.java

```diff
@@ -100,21 +100,49 @@ public static OptionalDouble mOr(OptionalDouble value1, OptionalDouble value2) {
   }
 
   public static OptionalDouble mAdd(OptionalDouble value1, OptionalDouble value2) {
-    return binopCondition((firstValue, secondValue) -> {
-      return OptionalDouble.of(firstValue.getAsDouble() + secondValue.getAsDouble());
-    }, value1, value2);
+
+    if(value1.isEmpty() && value2.isEmpty()) {
+      return OptionalDouble.empty();
+    }
+
+    double localValue1 = 0.;
+    double localValue2 = 0.;
```

Comment:
Wouahou that's painful. Maybe it's better to define a custom `MValue` class with a value field and an undefined field, that follows more closely what is done in `m_value.c` ?

Reply:
I can certainly transform `OptionalDouble` into a specific class but as I understand that would only partially solve the problem as you can't perform logical operations ( `||` or `&&` ) on integers in Java

Reply:
>  that would only partially solve the problem as you can't perform logical operations ( || or && ) on integers in Java

What is the problem with logical operations ? The way logical operations are defined in M follows what happen with bitwise operations (`|`, `&` in C), treating 0 as false and any non-zero as true.

Reply:
`x.value && y.value` does not work in Java, so you need a way to check if values are equal to 0.

For example, two integers toto and tata give this error message : 
```
bad operand types for binary operator '&&'
|    first type:  int
|    second type: int
|  tata && toto;
|  ^----------^
```

Reply:
Yep so you can do `x.value & y.value` to go the right result non ?

Reply:
In the m_value.c file I have it is && and ||, do I have a wrong version somehow ? 

Also using single &  doesn't seem to wotk either unfortunately : 

Values are 0 : 

```
jshell> tata = 0;
tata ==> 0

jshell> toto = 0;
toto ==> 0

jshell> tata & toto;
$17 ==> 0

```

Values are <> 0 :
```
jshell> toto = 10;
toto ==> 10

jshell> tata = 20;
tata ==> 20

jshell> tata & toto;
$20 ==> 0
```

Reply:
Ok here's the spec of `&&` and `||` in C: https://docs.microsoft.com/en-us/cpp/c-language/c-logical-operators?view=msvc-160. They match the semantics of `&` and `|` if and only if all the values manipulated are either 0 or 1. So it seems you would need a custom `m_and` and `m_or` function in Java to replicate the C semantics of `&&` and `||`, because we can't assume that M values that act as booleans are constrainted to either 0 or 1.

Reply:
I'm not sure I follow how my current (if inelegant) implementation should be changed to match those custom semantics

Reply:
You current implementation of `mAnd` and `mOr` are perfectly fine. My original comment concerned `mAdd` and `mSubstract` which have to use local variables.

Reply:
Oh of course, I beg your pardon, I got my paint brushes mixed up (to use a Frenchism), indeed a mvalue type similar to C would certainly give cleaner code. I'll put it down as a todo.

I've got the basic_case calculator working in Java so I think we're getting close to having a working java version, just few more details to iron out in the implementation and with the test harness.